evm/src/memory/memory_stark.rs

```diff
@@ -0,0 +1,551 @@
+        let next_values: Vec<_> = (0..8)
+            .map(|i| vars.next_values[sorted_value_limb(i)])
+            .collect();
+
+        let context_first_change = vars.local_values[CONTEXT_FIRST_CHANGE];
+        let segment_first_change = vars.local_values[SEGMENT_FIRST_CHANGE];
+        let virtual_first_change = vars.local_values[VIRTUAL_FIRST_CHANGE];
+        let address_unchanged =
+            one - context_first_change - segment_first_change - virtual_first_change;
+
+        let range_check = vars.local_values[RANGE_CHECK];
+
+        let not_context_first_change = one - context_first_change;
+        let not_segment_first_change = one - segment_first_change;
+        let not_virtual_first_change = one - virtual_first_change;
+        let not_address_unchanged = one - address_unchanged;
+
+        // First set of ordering constraint: first_change flags are boolean.
+        yield_constr.constraint(context_first_change * not_context_first_change);
+        yield_constr.constraint(segment_first_change * not_segment_first_change);
+        yield_constr.constraint(virtual_first_change * not_virtual_first_change);
+        yield_constr.constraint(address_unchanged * not_address_unchanged);
+
+        // Second set of ordering constraints: no change before the column corresponding to the nonzero first_change flag.
+        yield_constr
+            .constraint_transition(segment_first_change * (next_addr_context - addr_context));
+        yield_constr
+            .constraint_transition(virtual_first_change * (next_addr_context - addr_context));
+        yield_constr
+            .constraint_transition(virtual_first_change * (next_addr_segment - addr_segment));
+        yield_constr.constraint_transition(address_unchanged * (next_addr_context - addr_context));
+        yield_constr.constraint_transition(address_unchanged * (next_addr_segment - addr_segment));
+        yield_constr.constraint_transition(address_unchanged * (next_addr_virtual - addr_virtual));
+
+        // Third set of ordering constraints: range-check difference in the column that should be increasing.
+        let range_check_value = context_first_change * (next_addr_context - addr_context - one)
+            + segment_first_change * (next_addr_segment - addr_segment - one)
+            + virtual_first_change * (next_addr_virtual - addr_virtual - one)
+            + address_unchanged * (next_timestamp - timestamp - one);
+        yield_constr.constraint_transition(range_check - range_check_value);
+
+        // Enumerate purportedly-ordered log.
+        for i in 0..8 {
+            yield_constr
+                .constraint(next_is_read * address_unchanged * (next_values[i] - values[i]));
+        }
+
+        // Lookup argument for the range check.
+        let local_perm_input = vars.local_values[RANGE_CHECK_PERMUTED];
+        let next_perm_table = vars.next_values[COUNTER_PERMUTED];
+        let next_perm_input = vars.next_values[COUNTER_PERMUTED];
```

Comment:
Is it normal that these two values are the same?

Reply:
Oh right, the second is supposed to be `RANGE_CHECK_PERMUTED`. Good catch